statistics/sample_serial_test.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statistics
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/pingcap/tidb/util/mock"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"github.com/stretchr/testify/require"
+)
+
+type testSampleSuite struct {
+	count int
+	rs    sqlexec.RecordSet
+}
+
+func TestSample(t *testing.T) {
+	s := createTestSampleSuite()
+	t.Run("SubTestCollectColumnStats", SubTestCollectColumnStats(s))
```

Comment:
Returning closure make codes a little harder to understand. How about using [suite package](https://github.com/stretchr/testify#suite-package) of testify ? It also helps to avoid manually calling test case functions.

Reply:
hi, thanks for review. as per guide in https://github.com/pingcap/tidb/issues/26022, we did not use suite feature of testify.

Reply:
OK. I'm sorry that I missed the guide.

But on the other hand, I think the reason of avoid using `suite` is because suite runs serially inside (see [here](https://github.com/pingcap/tidb/pull/26082#issuecomment-877230972)).

In the case here, `SubTestCollectColumnStats`, `SubTestMergeSampleCollector`, and `SubTestCollectorProtoConversion` should run serially (see [here](https://github.com/pingcap/tidb/pull/28537/files#r722081163)), and would be suitable for using `suite`, with the benefit of avoid closures and manually calling test functions.

At the same time, `SubTestWeightedSampling`, `SubTestWeightedSampling`, and `SubTestBuildStatsOnRowSample` can run in parallel, and do not depend on `testSampleSuite` actually. It would be suitable to be individual test functions.

@tisonkun How do you think ?

Reply:
I tend to avoid suite. It is an abstraction lack of interoperates with Parallel test. You can run subtests in

* CLI by `go test -test.fun TestSample/SubTestCollectColumnStats`
* GoLand UI by click the subtest statements.

We don't have to use suite.

Reply:
Another discussion can be found at https://github.com/pingcap/tidb/issues/27178#issuecomment-921842586

Reply:
Got it, thanks @tisonkun . And looking forward to your wrapping ^_^

@feitian124 Please ignore this comment.